File: Dynamic Programming/213_house_robber_II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        dp1 = [-1] * n
        dp2 = [-1] * n

        def helper(nums, i, n, dp):
            if i >= n:
                return 0
            elif dp[i] != -1:
                return dp[i]

            dp[i] = max(nums[i] + helper(nums, i + 2, n, dp), helper(nums, i + 1, n, dp))
            return dp[i]

        return max(helper(nums, 0, n - 1, dp1), helper(nums, 1, n, dp2))

File: Dynamic Programming/213_house_robber_II.cpp
class Solution {
private:
    int helper(vector<int>& nums, int left, int right) {
        int one = 0, two = nums[left];
        for(int i = left + 1; i < right; i += 1) {
            int temp = two;
            two = max(one + nums[i], two);
            one = temp;
        }

        return two;
    }
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) return nums[0];
        return max(helper(nums, 0, n - 1), helper(nums, 1, n));
    }
};

// revised - 01/28/2025